### PR TITLE
Improve SettingsPanelPage animation

### PR DESCRIPTION
**Pull request by stonko1994**
**Issue: #224**

## Problem Description
_This PR actually solves two issues_
1) The calculation of the size of the `SettingsPanel` which is needed for the animation between `SettingPanelPages` lead to the issue that the `SettingsPanel` expands unexpectedly in certain cases.
2) If you have more than two pages the navigation was not working correctly as we were stuck in the penultimate page and were not able to navigate further back.

## Solution
1) The size calculation was rewritten and now only considers the size of the target `SettingsPanelPage`.
2) Don't push pages to the stack if we are navigating back.

## Tests
_Since we don't have UI tests (yet) the actual issue is not covered with tests_
- A simple test for navigation between pages was added.

## Manuel Tests
- To test the root issue you can select `UIWithSeparateAudioSubtitlesButtons` from the UI select box and click multiple times on the `subtitle` or `audio` button.
- If you want to test the animation a bit more you could use following UIContainer (Items and code are not 100% perfect but the animation should be the focus)
<details>

```ts
  let subtitleOverlay = new SubtitleOverlay();

    let subtitleListBox = new SubtitleListBox();
    let subtitleSettingsPanelPage1 = new SettingsPanelPage({
      components: [
        new SettingsPanelItem(null, subtitleListBox),
      ],
    });

    let mainSettingsPanelPage = new SettingsPanelPage({
      components: [
        new SettingsPanelItem('Video Quality', new VideoQualitySelectBox()),
        new SettingsPanelItem('Speed', new PlaybackSpeedSelectBox()),
        new SettingsPanelItem('Audio Track', new AudioTrackSelectBox()),
        new SettingsPanelItem('Audio Quality', new AudioQualitySelectBox()),
      ],
    });

    let settingsPanel = new SettingsPanel({
      components: [],
      hidden: true,
    });


    subtitleSettingsPanelPage1.addComponent(new SettingsPanelPageBackButton({
      container: settingsPanel,
      text: 'Back',
    }));

    let subtitleSettingsPanelPage = new SubtitleSettingsPanelPage({
      settingsPanel: settingsPanel,
      overlay: subtitleOverlay,
    });

    subtitleSettingsPanelPage1.addComponent(new SettingsPanelPageOpenButton({
      targetPage: subtitleSettingsPanelPage,
      container: settingsPanel,
    }));

    let subtitleSettingsOpenButton = new SettingsPanelPageOpenButton({
      targetPage: subtitleSettingsPanelPage1,
      container: settingsPanel,
    });

    mainSettingsPanelPage.addComponent(
      new SettingsPanelItem(
        new SubtitleSettingsLabel({ text: 'Subtitles', opener: subtitleSettingsOpenButton }),
        new SubtitleSelectBox()
      ));

    settingsPanel.addComponent(mainSettingsPanelPage);
    settingsPanel.addComponent(subtitleSettingsPanelPage1);
    settingsPanel.addComponent(subtitleSettingsPanelPage);

    let controlBar = new ControlBar({
      components: [
        settingsPanel,
        new Container({
          components: [
            new PlaybackTimeLabel({ timeLabelMode: PlaybackTimeLabelMode.CurrentTime, hideInLivePlayback: true }),
            new SeekBar({ label: new SeekBarLabel() }),
            new PlaybackTimeLabel({ timeLabelMode: PlaybackTimeLabelMode.TotalTime, cssClasses: ['text-right'] }),
          ],
          cssClasses: ['controlbar-top'],
        }),
        new Container({
          components: [
            new PlaybackToggleButton(),
            new VolumeToggleButton(),
            new VolumeSlider(),
            new Spacer(),
            new PictureInPictureToggleButton(),
            new AirPlayToggleButton(),
            new CastToggleButton(),
            new VRToggleButton(),
            new SettingsToggleButton({ settingsPanel: settingsPanel }),
            new FullscreenToggleButton(),
          ],
          cssClasses: ['controlbar-bottom'],
        }),
      ],
    });

    return new UIContainer({
      components: [
        subtitleOverlay,
        new BufferingOverlay(),
        new PlaybackToggleOverlay(),
        new CastStatusOverlay(),
        controlBar,
        new TitleBar(),
        new RecommendationOverlay(),
        new Watermark(),
        new ErrorMessageOverlay(),
      ],
    });
```
</details>

### CODE REVIEW

File: src/ts/components/settingspanel.ts [hunk 1]
```diff
@@ -25,15 +26,40 @@ export interface SettingsPanelConfig extends ContainerConfig {
   pageTransitionAnimation?: boolean;
 }
 
+enum NavigationDirection {
+  Forward,
+  Backwards,
```
Comment (protyposis) [inline src/ts/components/settingspanel.ts]: `Backward` (either Forward/Backward or Forwards/Backwards)

File: src/ts/components/settingspanel.ts [hunk 2]
```diff
@@ -93,78 +121,157 @@ export class SettingsPanel extends Container<SettingsPanelConfig> {
...
+   * Returns the current active / visible page
+   * @return {SettingsPanelPage}
+   */
+  getActivePage(): SettingsPanelPage {
+    return this.activePage;
   }
 
+  /**
+   * Sets the
+   * @deprecated Use setActivePage instead
```
Comment (protyposis) [inline src/ts/components/settingspanel.ts]: `{@link setActivePage}`

File: src/ts/components/settingspanel.ts [hunk 3]
```diff
@@ -93,78 +121,157 @@ export class SettingsPanel extends Container<SettingsPanelConfig> {
...
-    }
+    this.setActivePage(this.getPages()[index]);
   }
 
-  setActivePage(page: SettingsPanelPage): void {
-    const index = this.getPages().indexOf(page);
-    this.setActivePageIndex(index);
+  /**
+   * Adds the passed page to the navigation stack and make it visible. Do not use this method to navigate backwards!
+   * Results in no-op if the target pate is the current page.
```
Comment (protyposis) [inline src/ts/components/settingspanel.ts]: Typo

File: src/ts/components/settingspanel.ts [hunk 4]
```diff
@@ -93,78 +121,157 @@ export class SettingsPanel extends Container<SettingsPanelConfig> {
...
-      targetPage.onActiveEvent();
-    }
+    this.setActivePage(this.getPages()[index]);
   }
 
-  setActivePage(page: SettingsPanelPage): void {
-    const index = this.getPages().indexOf(page);
-    this.setActivePageIndex(index);
+  /**
+   * Adds the passed page to the navigation stack and make it visible. Do not use this method to navigate backwards!
```
Comment (protyposis) [inline src/ts/components/settingspanel.ts]: Typo: make => makes

Comment (protyposis) [reply src/ts/components/settingspanel.ts]: Maybe add a `use {@link popSettingsPanelPage}` to navigate backwards hint

File: src/ts/components/settingspanel.ts [hunk 5]
```diff
@@ -93,78 +121,157 @@ export class SettingsPanel extends Container<SettingsPanelConfig> {
...
+    if (targetPage === this.getActivePage()) {
+      console.warn('Page is already the current one ... skipping navigation');
+      return;
+    }
+
+    this.navigateToPage(targetPage, this.getActivePage(), NavigationDirection.Forward);
   }
 
+  /**
+   * Resets the navigation stack and navigate to the root page and make it visible
```
Comment (protyposis) [inline src/ts/components/settingspanel.ts]: "Resets the navigtion stack by navigating back to the root page and displaying it."

File: src/ts/components/settingspanel.ts [hunk 6]
```diff
@@ -93,78 +121,157 @@ export class SettingsPanel extends Container<SettingsPanelConfig> {
...
+   */
   popToRootSettingsPanelPage(): void {
     this.resetNavigation();
   }
 
+  /**
+   * Removes the current page form the navigation stack and make the previous one visible.
```
Comment (protyposis) [inline src/ts/components/settingspanel.ts]: form => from, make => makes

File: src/ts/components/settingspanel.ts [hunk 7]
```diff
@@ -93,78 +121,157 @@ export class SettingsPanel extends Container<SettingsPanelConfig> {
...
+   * Results in a no-op if we are already on the rootPage.
```
Comment (protyposis) [inline src/ts/components/settingspanel.ts]: "root page"

**Comment by stonko1994** [comment]
@protyposis I addressed the issue you pointed out in 2bc6598. There is also a test covering this.

**Pull request merged**